clang/lib/Headers/cheriintrin.h

```diff
@@ -0,0 +1,132 @@
+#define cheri_gettag(x) __builtin_cheri_tag_get(x)
+#define cheri_cleartag(x) __builtin_cheri_tag_clear(x)
+#define cheri_is_subset(x, y) __builtin_cheri_subset_test((x), (y))
+
+/* Bounds setting intrinsics: */
+#define cheri_round_representable_length(x)                                    \
+  __builtin_cheri_round_representable_length(x)
+#define cheri_representable_alignment_mask(x)                                  \
+  __builtin_cheri_representable_alignment_mask(x)
+#define cheri_setbounds(x, y) __builtin_cheri_bounds_set((x), (y))
+#define cheri_setbounds_exact(x, y) __builtin_cheri_bounds_set_exact((x), (y))
+
+/* Object types, sealing and unsealing: */
+typedef long cheri_otype_t;
+#if defined(__mips__) || defined(__riscv)
+/* CHERI-MIPS and CHERI-RISCV use negative numbers for hardware-interpreted
+ * otypes */
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)-1)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)-2)
+#else
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)0)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)1)
+#endif
+#define cheri_gettype(x) __builtin_cheri_type_get(x)
+#define cheri_is_sealed(x) __builtin_cheri_sealed_get(x)
+#define cheri_is_sentry(x) (__builtin_cheri_type_get(x) == CHERI_OTYPE_SENTRY)
+#define cheri_is_unsealed(x) (!__builtin_cheri_sealed_get(x))
+/* TODO: builtins for indirect sentries */
+#define cheri_seal_entry(x) __builtin_cheri_seal_entry(x)
+#define cheri_seal(x, y) __builtin_cheri_seal((x), (y))
+#define cheri_unseal(x, y) __builtin_cheri_unseal((x), (y))
+
+/* Reconstruct capabilities from raw data: */
+#define cheri_build_cap(x, y) __builtin_cheri_cap_build((x), (y))
+#define cheri_conditional_seal(x, y) __builtin_cheri_conditional_seal((x), (y))
+#define cheri_copy_type(x, y) __builtin_cheri_cap_type_copy((x), (y))
+
+/* Capability permissions: */
+enum __attribute__((flag_enum, enum_extensibility(open))) CheriPermissions {
+  CHERI_PERM_GLOBAL = __CHERI_CAP_PERMISSION_GLOBAL__,
+  CHERI_PERM_EXECUTE = __CHERI_CAP_PERMISSION_PERMIT_EXECUTE__,
+  CHERI_PERM_LOAD = __CHERI_CAP_PERMISSION_PERMIT_LOAD__,
+  CHERI_PERM_STORE = __CHERI_CAP_PERMISSION_PERMIT_STORE__,
+  CHERI_PERM_LOAD_CAP = __CHERI_CAP_PERMISSION_PERMIT_LOAD_CAPABILITY__,
+  CHERI_PERM_STORE_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_CAPABILITY__,
+  CHERI_PERM_STORE_LOCAL_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_LOCAL__,
+  CHERI_PERM_SEAL = __CHERI_CAP_PERMISSION_PERMIT_SEAL__,
+  CHERI_PERM_CCALL = __CHERI_CAP_PERMISSION_PERMIT_CCALL__,
+  CHERI_PERM_UNSEAL = __CHERI_CAP_PERMISSION_PERMIT_UNSEAL__,
+  CHERI_PERM_SYSTEM_REGS = __CHERI_CAP_PERMISSION_ACCESS_SYSTEM_REGISTERS__,
+  /* TODO: architecture-dependent permissions */
+};
+typedef enum CheriPermissions cheri_perms_t;
+#define cheri_getperms(x) ((cheri_perms_t)__builtin_cheri_perms_get(x))
+#define cheri_andperms(x, y) __builtin_cheri_perms_and((x), (cheri_perms_t)(y))
+#define cheri_clearperms(x, y) cheri_andperms((x), ~(y))
```

Comment:
Need to be very careful here strictly speaking and cast to `__SIZE_TYPE__` before doing the bitwise negation. In practice none of our architectures make use of the full architectural 64-bit permissions field as there aren't enough bits in the encoding, but we should be as agnostic to that as possible.